### PR TITLE
fix: ensure that per-element investment cost are calculated accurately

### PR DESCRIPTION
**Pull request by danielolsen**
### Purpose
- As originally mentioned in https://github.com/Breakthrough-Energy/PowerSimData/pull/297, fix AC investment cost calculations so that we don't use underground line costs when not appropriate (lines of appropriate rating.voltage that are entirely within `NEISO` and `NYISO J-K` regions).
- Add tests of individual element cost calculations, which illustrate some bugs (e.g. branches within the `SOCO` [Southern Company] region were getting `NaN` cost estimates, and underground line costs were being applied inappropriately to some branches).
- Fix the bugs.
- Cleanup from https://github.com/Breakthrough-Energy/PowerSimData/pull/391, where we enabled the cost calculation of nuclear generators, but didn't ensure that costs would be calculated for costs cases other than `"Moderate"`.

These code fixes are necessary for https://github.com/Breakthrough-Energy/PowerSimData/issues/439.

### What the code is doing
- In **powersimdata/tests/mock_grid.py**: we allow the user to pass data to the `MockGrid.storage["StorageData"]` field for tests.
- In **powersimdata/design/investment/tests/test_investment_costs.py**: we add tests of the individual element cost calculations (which initially fail)
- In **powersimdata/design/investment/investment_costs.py**:
  - We add a new function `merge_keep_index`, which lets us perform a dataframe merge without losing the index of the first dataframe (as demonstrated in the comments on https://github.com/Breakthrough-Energy/PowerSimData/issues/439), and we use this wherever we had been doing a `pandas.DataFrame.merge`.
  - We add a new function `append_keep_index_name` which lets us perform an append without losing the name of the first dataframe index, which we need for `merge_keep_index`.
  - In `_calculate_ac_inv_costs`:
    - We update the `select_mw` function to ensure that we only select the cost class for underground lines if the line is entirely within one of the two zones which have significant shares of underground lines (based on the data in the **LineRegMult.csv** file).
    - We generalize the `get_transformer_mult` function to be `get_branch_mult`, so that it can operate on all types of branches to return the average of the regional multiplier for the to & from ends of the branch. In the process, we fix a bug where if a Line has endpoints with no regional multiplier defined, it would return a `NaN` multiplier and then this branch would not count towards the total cost (and not be meaningful when results are returned un-summed).
    - Adding data to the branch data frame is moved after adding data to the bus data frame, so that we can take advantage of knowing the bus regions within the new `get_branch_mult` function.
  - In `calculate_gen_inv_costs`:
    - We modify the `load_cost` function so that if we request a cost case that is not `"Moderate"`, then we copy the `"Moderate"` nuclear costs into what we will return, and warn the user. This allows us to calculate costs for the `"Advanced"` and `"Conservative"` cost cases without having `NaN`s for nuclear.
    - We ensure that the indices of the storage units that we append to the plant data frame are preserved, so that there is no collision with the plant IDs in the returned non-summed data frame. We generate these non-colliding indices automatically whenever we add storage to the Grid, but they live in the `StorageData` storage table due to legacy design around the Matpower/MOST schema)

### Testing
Unit tests are added which detect the bugs, and the changes fix the bugs so that the tests run successfully. This branch has also been tested (with some additional code in the `daniel/congestion_cost_upgrade` branch) to successfully analyze per-branch upgrade costs for a real Eastern scenario.

### Time estimate
30-60 minutes. The `investment_cost` module is big, and requires aggregating many different tables to produce results, so it takes a while to trace through all the places where bugs can be introduced. There is definitely room to refactor this module to be structured a bit better, but that is of lower priority than fixing these bugs, which need to be addressed for the trainings that are coming up soon.

### CODE REVIEW

File: powersimdata/design/investment/investment_costs.py [hunk 1]
```diff
@@ -13,6 +14,30 @@
 from powersimdata.utility.distance import haversine
 
 
+def merge_keep_index(df1, df2, **kwargs):
+    """Execute a pandas DataFrame merge, preserving the index of the first dataframe.
+
+    :param pandas.DataFrame df1: first data frame, to call pandas merge from.
+    :param pandas.DataFrame df2: second data frame, argument to pandas merge.
+    :param \\*\\*kwargs: arbitrary keyword arguments passed to pandas merge call.
```
Comment (BainanXia) [inline powersimdata/design/investment/investment_costs.py:22]: Missing `return` docstrings.

File: powersimdata/design/investment/investment_costs.py [hunk 2]
```diff
@@ -13,6 +14,30 @@
...
+    return df1.reset_index().merge(df2, **kwargs).set_index(df1.index.names)
+
+
+def append_keep_index_name(df1, other, *args, **kwargs):
+    """Execute a pandas DataFrame append, preserving the index name of the dataframe.
+
+    :param pandas.DataFrame df1: first data frame, to call pandas append from.
+    :param pandas.DataFrame/pandas.Series/list: first argument to pandas append method.
+    :param \\*args: arbitrary positional arguments passed to pandas append call.
+    :param \\*\\*kwargs: arbitrary keyword arguments passed to pandas append call.
```
Comment (BainanXia) [inline powersimdata/design/investment/investment_costs.py:33]: Same here.

File: powersimdata/design/investment/investment_costs.py [hunk 3]
```diff
@@ -59,27 +84,35 @@ def select_mw(x, cost_df):
         :return: (*pandas.Series*) -- series of [*'MW'*, *'costMWmi'*] to be assigned
             to branch.
         """
-
+        underground_regions = ("NEISO", "NYISO J-K")
+        cost_df = cost_df.copy()
```
Comment (BainanXia) [inline powersimdata/design/investment/investment_costs.py:88]: I understand the purpose here, however, it would be better to use a different variable name to reduce confusion.

File: powersimdata/design/investment/investment_costs.py [hunk 4]
```diff
@@ -59,27 +84,35 @@ def select_mw(x, cost_df):
...
         # get rid of NaN values in this kV table
-        tmp = tmp[~tmp["MW"].isna()]
+        cost_df = cost_df[~cost_df["MW"].isna()]
         # find closest MW & corresponding cost
-        return tmp.iloc[np.argmin(np.abs(tmp["MW"] - x.rateA))][["MW", "costMWmi"]]
+        return_cols = ["MW", "costMWmi"]
+        return cost_df.iloc[np.argmin(np.abs(cost_df["MW"] - x.rateA))][return_cols]
 
-    def get_transformer_mult(x, bus_reg, ac_reg_mult, xfmr_lookup_alerted=set()):
+    def get_branch_mult(x, bus_reg, ac_reg_mult, branch_lookup_alerted=set()):
```
Comment (BainanXia) [inline powersimdata/design/investment/investment_costs.py:100]: I think based on our convention, we should avoid mutable default values?

File: powersimdata/design/investment/investment_costs.py [hunk 5]
```diff
@@ -155,21 +170,36 @@ def get_transformer_mult(x, bus_reg, ac_reg_mult, xfmr_lookup_alerted=set()):
 
     bus_reg.drop(["lat", "lon"], axis=1, inplace=True)
 
+    # Add extra information to branch data frame
+    branch = grid_new.branch
+    branch.loc[:, "kV"] = branch.apply(
+        lambda x: bus.loc[x.from_bus_id, "baseKV"], axis=1
+    )
```
Comment (BainanXia) [inline powersimdata/design/investment/investment_costs.py:177]: Should this be equivalent to `branch["kV"] = bus.loc[branch["from_bus_id"], "baseKV"]`?

File: powersimdata/design/investment/investment_costs.py (hunk 4 above)
Comment (danielolsen) [reply powersimdata/design/investment/investment_costs.py:100]: This design is intentional. Since we re-define `get_branch_mult` for every call of `_calculate_ac_inv_costs`, it is never shared via user-accessible calls. We do want it to be shared via the many calls to `get_branch_mult` via the dataframe `apply`, so that it serves its purpose of not the user multiple times for the same missing regional multipliers.

I believe this was discussed at some point in the review for https://github.com/Breakthrough-Energy/PowerSimData/pull/297.

File: powersimdata/tests/mock_grid.py [hunk 6]
```diff
@@ -151,6 +151,22 @@
...
+        "TerminalStoragePrice",
+        "MinStorageLevel",
+        "MaxStorageLevel",
+        "OutEff",
+        "InEff",
+        "LossFactor",
+        "rho",
+        "ExpectedTerminalStorageMax",
+        "ExpectedTerminalStorageMin",
+    ],
```
Comment (BainanXia) [inline powersimdata/tests/mock_grid.py:169]: Nice.

File: powersimdata/tests/mock_grid.py [hunk 7]
```diff
@@ -14,7 +14,7 @@
 }
 
 gencost_names = {"gencost_before": "before", "gencost_after": "after"}
-storage_names = {"storage_gen": "gen"}
+storage_names = {"storage_gen": "gen", "storage_StorageData": "StorageData"}
```
Comment (BainanXia) [inline powersimdata/tests/mock_grid.py:17]: Shall we use "storage_data" instead of "storage_StorageData" or there is another `storage_data` defined somewhere else.

File: powersimdata/design/investment/investment_costs.py (hunk 4 above)
Comment (BainanXia) [reply powersimdata/design/investment/investment_costs.py:100]: Nevermind then.

**Comment by BainanXia** [comment]
Thanks for updating the tests and the Mock object. The logic is clear and all tests passed. I think this is good to go after some minor concerns are addressed.

File: powersimdata/tests/mock_grid.py (hunk 7 above)
Comment (danielolsen) [reply powersimdata/tests/mock_grid.py:17]: The logic here was: `"storage_gen"` reflects the `"gen"` key in the `storage` dict. Therefore, `"storage_StorageData"` reflects the `"StorageData"` key in the `storage` dict. But it is arbitrary, I don't mind changing it to whatever you prefer.

There's a longer term project here to update our storage schema to something more intuitive (now that we don't need to maintain compatibility with Matpower), but that will require deeper design thinking.

**Comment by danielolsen** [comment]
I found another edge case that also needs fixing: if there are no transformers in the upgraded branches then we get an error caused by trying to `apply` on an empty data frame. I will add some tests to detect the problem and a fix for it, and let you know when they're ready.

File: powersimdata/tests/mock_grid.py (hunk 7 above)
Comment (BainanXia) [reply powersimdata/tests/mock_grid.py:17]: I see. Either way. No strong preference ai this point.

File: powersimdata/design/investment/investment_costs.py (hunk 5 above)
Comment (danielolsen) [reply powersimdata/design/investment/investment_costs.py]: Done (simplified this one and the two after it).

File: powersimdata/design/investment/investment_costs.py (hunk 3 above)
Comment (danielolsen) [reply powersimdata/design/investment/investment_costs.py]: Done.

File: powersimdata/design/investment/investment_costs.py (hunk 2 above)
Comment (danielolsen) [reply powersimdata/design/investment/investment_costs.py:34]: Done.

File: powersimdata/design/investment/investment_costs.py (hunk 1 above)
Comment (danielolsen) [reply powersimdata/design/investment/investment_costs.py:22]: Done.

**Comment by danielolsen** [comment]
> 
> 
> I found another edge case that also needs fixing: if there are no transformers in the upgraded branches then we get an error caused by trying to `apply` on an empty data frame. I will add some tests to detect the problem and a fix for it, and let you know when they're ready.

New tests and fixes are added, and all comments are addressed (I think).

**Review by BainanXia — APPROVED** [review]
Looks good. Thanks!
